src/modules/mc_pos_control/mc_pos_control_main.cpp

```diff
@@ -640,8 +578,26 @@ MulticopterPositionControl::Run()
 				_flight_tasks.reActivate();
 			}
 
-			// Run position control
-			_control.setState(_states);
+			// Update position control
+			PositionControlStates states {
+				.position = Vector3f{local_pos.x, local_pos.y, local_pos.z},
+				.velocity = velocity,
+				.acceleration = acceleration,
+				.yaw = local_pos.heading,
+			};
+
+			// If a change in velocity is demanded set velocity to the derivative of position
+			// because it has less bias but blend it in across the landing speed range
+			if (PX4_ISFINITE(setpoint.vz) && local_pos.z_valid) {
+				const float vel_sp_z_abs = fabsf(setpoint.vz);
+
+				if (vel_sp_z_abs > FLT_EPSILON) {
+					const float weighting = fminf(vel_sp_z_abs / _param_mpc_land_speed.get(), 1.f);
+					states.velocity(2) = local_pos.z_deriv * weighting + local_pos.vz * (1.f - weighting);
```

Comment:
I guess, the intent is to use the position derivative that has less bias when the demanded velocity is small to avoid that weird behavior of the drone not moving or going in the wrong direction. However, the weighting has the opposite effect.

Reply:
Yes, I was surprised by this implementation as well, and went back to verify the original logic.

Reply:
I vaguely recall this being done in response to a flyaway years ago.

Reply:
This still seems backwards to me, let's review on the dev call?

Reply:
Yes, there were situations where the vehicles were flying slowly upwards with a commanded downwards land speed. The goal is to have slow vertical velocities listen to the position derivative because of the possible small biases on the velocity estimate but using the more accelerometer-based velocity estimate for good hover peformance. It's by Dennis and we had multiple discussions about it with @priseborough .

Reply:
The weighting represents vertical velocity setpoint:
- 0  -> use velocity estimate for good hover performance (@priseborough's argument for the high reliance on accelerometer integration)
- < `MPC_LAND_SPEED` ramp up using altitude derivative without a step
- >= `MPC_LAND_SPEED` use altitude derivative